README.txt
=== ALU WMS Metrocell Power Menu ===
|     Creator: rendicott@gmail.com
|     Date: 2014-09-15
|     Version: 0.4.7
====================================

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~CONTENTS~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-INTRODUCTION
-PREREQS / INSTALLATION
-USAGE
-ALARM DB INTERACTIONS
-INVENTORY DB INTERACTIONS
-OTHER COMMANDS
-REGULAR OPERATION
-ADVANCED INFORMATION
-APPENDIX


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~INTRODUCTION~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This tool was created as a way to speed up the monitoring and administration of the Alcatel-Lucent Small Cell solution through WMS/HDM. This tool cannot do 
everything that WMS and HDM can do but it is a very useful supplement. 

This document is to help get the user started on using the tool and to outline some of the capabilities and rules for usage. 


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PREREQS / INSTALLATION~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This tool is designed to be used from the SSH command prompt of the WMS Server. At the time of this writing that means using Putty to SSH to WMS like in 
the below example:

	ssh USERNAME@METROCELL-ALP2WMS-CLIENT nesa-aln.web.bc.com:8022

To obtain a copy of the PowerMenu perform the following steps:

Navigate to your home directory on WMS:

        $ cd ~

Create a new directory. Here we'll just call the directory 'pm' for "PowerMenu"

        $ mkdir pm

Navigate to the new directory

        $ cd pm

Unpack the latest PowerMenu release into the current directory:

        $ tar xvf /home/customer/russ/powermenu/powermenu-v0.4.7.tar
        x ./EXP-apcheck-new.exp, 2636 bytes, 6 tape blocks
        x ./EXP-apcheck.exp, 2432 bytes, 5 tape blocks
        x ./EXP-apreboot.exp, 718 bytes, 2 tape blocks
        x ./EXP-getbsg.exp, 373 bytes, 1 tape blocks
        x ./EXP-kh.exp, 324 bytes, 1 tape blocks
        x ./INSTRUCTIONS-UPGRADE.txt, 3606 bytes, 8 tape blocks
        x ./INSTRUCTIONS.txt, 3564 bytes, 7 tape blocks
        x ./PowerMenu.Job.Aid.docx, 2582839 bytes, 5045 tape blocks
        x ./README.txt, 192427 bytes, 376 tape blocks
        x ./RELEASENOTES.txt, 4811 bytes, 10 tape blocks
        x ./body.txt, 83 bytes, 1 tape blocks
        x ./pm.py, 61243 bytes, 120 tape blocks
        x ./pmClasses.py, 19777 bytes, 39 tape blocks
        x ./pmConfig.py, 18256 bytes, 36 tape blocks
        x ./pmFunctions.py, 43125 bytes, 85 tape blocks
        x ./pmLoaders.py, 76862 bytes, 151 tape blocks
        x ./pmRTest.py, 3054 bytes, 6 tape blocks
        x ./pmThreads.py, 26288 bytes, 52 tape blocks


Your directory contents should look like this:

        $ ls -l
        total 6010
        -rw-r--r--   1 russ   grouper     2636 Feb  9 13:48 EXP-apcheck-new.exp
        -rw-r--r--   1 russ   grouper     2432 Feb  9 13:48 EXP-apcheck.exp
        -rw-r--r--   1 russ   grouper      718 Feb  9 13:48 EXP-apreboot.exp
        -rw-r--r--   1 russ   grouper      373 Feb  9 13:48 EXP-getbsg.exp
        -rw-r--r--   1 russ   grouper      324 Feb  9 13:48 EXP-kh.exp
        -rw-r--r--   1 russ   grouper     3606 Feb  9 13:48 INSTRUCTIONS-UPGRADE.txt
        -rw-r--r--   1 russ   grouper     3564 Feb  9 13:48 INSTRUCTIONS.txt
        -rw-r--r--   1 russ   grouper  2582839 Feb  9 13:50 PowerMenu.Job.Aid.docx
        -rw-r--r--   1 russ   grouper   192427 Feb  9 13:48 README.txt
        -rw-r--r--   1 russ   grouper     4811 Feb  9 13:48 RELEASENOTES.txt
        -rw-r--r--   1 russ   grouper       83 Feb  9 13:00 body.txt
        -rw-r--r--   1 russ   grouper    61243 Feb  9 13:48 pm.py
        -rw-r--r--   1 russ   grouper    19777 Feb  9 13:48 pmClasses.py
        -rw-r--r--   1 russ   grouper    18256 Feb  9 13:48 pmConfig.py
        -rw-r--r--   1 russ   grouper    43125 Feb  9 13:48 pmFunctions.py
        -rw-r--r--   1 russ   grouper    76862 Feb  9 13:48 pmLoaders.py
        -rw-r--r--   1 russ   grouper     3054 Feb  9 13:48 pmRTest.py
        -rw-r--r--   1 russ   grouper    26288 Feb  9 13:48 pmThreads.py



Add execute permissions to the .exp scripts inside the directory.

        $ chmod +x *.exp

Now you should see that the .exp files have the execute permissions.


        $ ls -l
        total 6010
        -rwxr-xr-x   1 russ   grouper     2636 Feb  9 13:48 EXP-apcheck-new.exp
        -rwxr-xr-x   1 russ   grouper     2432 Feb  9 13:48 EXP-apcheck.exp
        -rwxr-xr-x   1 russ   grouper      718 Feb  9 13:48 EXP-apreboot.exp
        -rwxr-xr-x   1 russ   grouper      373 Feb  9 13:48 EXP-getbsg.exp
        -rwxr-xr-x   1 russ   grouper      324 Feb  9 13:48 EXP-kh.exp
        -rw-r--r--   1 russ   grouper     3606 Feb  9 13:48 INSTRUCTIONS-UPGRADE.txt
        -rw-r--r--   1 russ   grouper     3564 Feb  9 13:48 INSTRUCTIONS.txt
        -rw-r--r--   1 russ   grouper  2582839 Feb  9 13:50 PowerMenu.Job.Aid.docx
        -rw-r--r--   1 russ   grouper   192427 Feb  9 13:48 README.txt
        -rw-r--r--   1 russ   grouper     4811 Feb  9 13:48 RELEASENOTES.txt
        -rw-r--r--   1 russ   grouper       83 Feb  9 13:00 body.txt
        -rw-r--r--   1 russ   grouper    61243 Feb  9 13:48 pm.py
        -rw-r--r--   1 russ   grouper    19777 Feb  9 13:48 pmClasses.py
        -rw-r--r--   1 russ   grouper    18256 Feb  9 13:48 pmConfig.py
        -rw-r--r--   1 russ   grouper    43125 Feb  9 13:48 pmFunctions.py
        -rw-r--r--   1 russ   grouper    76862 Feb  9 13:48 pmLoaders.py
        -rw-r--r--   1 russ   grouper     3054 Feb  9 13:48 pmRTest.py
        -rw-r--r--   1 russ   grouper    26288 Feb  9 13:48 pmThreads.py



~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ USAGE ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~



Launch the script with basic functionality like this:

	python pm.py
	
At this point the script will start logging to a file named "pm.log" in the same directory. 

You'll see the following out put on the screen:

        $ python pm.py

         _____                       __  __
        |  __ \                     |  \/  |
        | |__) |____      _____ _ __| \  / | ___ _ __  _   _
        |  ___/ _ \ \ /\ / / _ \ '__| |\/| |/ _ \ '_ \| | | |
        | |  | (_) \ V  V /  __/ |  | |  | |  __/ | | | |_| |
        |_|   \___/ \_/\_/ \___|_|  |_|  |_|\___|_| |_|\__,_|
               === ALU WMS Metrocell Power Menu ===
               |     Creator: rendicott@gmail.com
               |     Date: 2014-11-20
               |     Version: 0.4.7
               ====================================


        Current Directory: /home/customer/russ/pm

        Testing color support...
        starting up with loglevel 50 CRITICAL
        newest file detected: ../pm_foodfiles/2015-02-09-06-31-05Powermenu-Food-CTS.csv

        Enter the password for the admin@FGW user:
        Enter again for verification:

        Enter the password for the old localOperator@AP user:
        Enter again for verification:

        Enter the password for the old localAdmin@AP user:
        Enter again for verification:

        Enter the password for the new 14.2 localOperator@AP user:
        Enter again for verification:

        Enter the password for the new 14.2 localAdmin@AP user:
        Enter again for verification:

The program will ask you for the five passwords required in order to interact with the Metrocells and FemtoGateways out on the network. 
You'll have to enter each password twice upon launching the program. Alternatively you can pass the passwords in as parameters. 

	|---------------------------------------------------------------------------------------
	|
	| PASSING PASSWORDS AS PARAMETERS INSTEAD OF ENTERING INTERACTIVELY:

RELEASENOTES.txt
Known Bugs:
-when launching script with the '-d' option if you pass a '-l' loglevel instead of '-1' the script bombs
-Bug in alarm info when BSRName deviates from XXXXXXXNNNNN format (e.g., spaces in name for lab ap's)
-Fix up regex matching, example: typing "exit" tries to send email
-with 2+index, alarm timestamps not displaying properly
-put validation into food csv
-UNVERIFIED: bad password handling
-In alarm view (9+XX) IuPS etc alarms show up as adminStatus 'locked' which is wrong.
-add better TIMEOUT handling to reboot sequence
-fix date sort order in cell up/down log

REQUESTS:
-Ignore 'lmcFailure' alarms completely. 

Known Limitations:
-The usefulness of the script depends upon having fresh data. Fresh data is awkward to obtain for most users.
--Want to write in a "if db older than 4 hrs force refresh"
--Future development goals include behind-the-scenes data refreshes.
-The WICL 'FM ClearAlarm' command does not work. There is already an AR open with ALU (AR1-5424256)
--If the clearalarm command worked the tool could be much more robust.

<<<<<<< HEAD
v0.4.6.a
------
- fixed bug where AP checks sometimes timout
=======
v0.4.7
------
-cleaned up a ton of formatting, code is closer to being PEP8 compliant
-added threading but due to WMS performance issues, limiting to a single thread at a time
-added AP software version display on metrocell fullformat view
-added support for new 14.02 passwords, Powermenu now needs five passwords fed into it
-added AP syslog collection-zip-and-email function for individual and group queries.
-added custom query builder
-added '114' option to refresh registrations and alarms in one command
-added MAC address pull to AP check
-added cluster alarm 'csCnConnectivityLost' to stuck alarm check

>>>>>>> rolling

v0.4.6
------
-Switched from 'unmonfiles' to 'CTS food' files. Auto detects and loads latest from '/home/customer/russ/pm_foodfiles'
---Food report is sourced from AOTS and it's a CSV of the "Daily Small Cell Breakdown" CTS_rawdata tab. 
---added "full" format which shows more info for inventory view like USID and FALOC code (toggle with 'f')
-Added 8 second (configurable) timeout to AP checks
-removed stored passwords in expect scripts. User is now prompted for all three passwords or can pass in as parameters.
--Added LCell up/down history log for quick "Return to service" time reference. Shows in AP check "verbose" results.\
-Added 3+27,3+28, and 3+29 to search inventory by CtsEquipName, USID, and FALocation codes

v0.4.5.a
------
-Fixed false negatives bug with bad 'funcname' logging. 

v0.4.5
------
-fixed status check "0 of 10", now starts at "1 of 10"
-added '113' full refresh to help
-status check results now display cluster/group to make it easier to run follow up checks on group
-inventory view for an AP now shows an asterisk next to AP if it has any alarms
--alarms for that AP can be looked up with the 2+index command
-All WMS data now comes from flat files to limit WICL sessions.
--by default the files dump to /home/customer/russ/wicldumps
-GRID/GATEWAY info fixed. Parsing it takes longer though. (see option -g)
-Added 'Santa Ana 2' as new monitored FGW (SNAPCAHHCR9S02)
-fixed false negatives known_hosts SSH check
-fixed 113/112 refresh updated
-fixed: if launch menu.py without unmonfile and try to 113 refresh, will bomb.
-Added AP reboot feature (6+commonid). 

v0.4.4
------
-fixed colors bug with CLI*Manager

v0.4.3
-------
-general improvements and bug fixes
-added CSV output and CSV-to-Email function
-added Stuck alarms check for BSR related alarms
-began conversion from Alarm dictionary to Alarm Objects (LOO = List of Objects)
-added macro functionality


v0.4.1
-------
-added bulk CommonID search with 'b' and list from csv string

v0.4
-------
-added bulk ap checks based on cluster,group search
-added rowstatus display for ap check results
-began elimniating global variables and instead pass relevant variables from function to function and use returns
-added error handling for the bug when ssh keys don't match known_hosts for user profile
-added a lot more debugging strings
-added basic README and RELEASENOTES
-added timestamps to the results display tables in preparation for v0.5 with background data updates
-cleaned up a lot of formatting and made it readable on smaller screens
-added feature to show a list of AP's that updated with '112'

v0.3
-------
fixed minor bugs
added adminStatus to alarm view
added color coding to adminStatus and ipsecIpAddr
added SuperList subclass
added logging and debugging


Next in 0.4.4 and beyond (Hopefully)

pmConfig.py

'''
---ALU Metrocell WMS PowerMenu-----
# This program makes it easier to run
# diagnostic commands from the NESA
# ALU WMS. 
#
# Created by Russell Endicott (rendicott@gmail.com)
# 2014-08-19
'''

'''
This PowerMenu module is a place to store static variable input data.
Eventually stuff in this module should be stored in a database or user
configurable file. For now I'm just collecting it here so there's very little
content in the actual code that's 'hard coded'. 
'''
sversion = '0.4.8'


#some hardcoded script names for the two BASH scripts, need to do this better
''' NOTE: I had to put the /usr/bin/expect in front of the script 
name because WMS shell (ksh) is silly'''
fgw_bsgScript = '/usr/bin/expect ./EXP-getbsg.exp'
femtostatus_script = '/usr/bin/expect ./EXP-apcheck.exp'
femtostatus_script_new = '/usr/bin/expect ./EXP-apcheck-new.exp'
knownhosts_script = '/usr/bin/expect ./EXP-kh.exp'
reboot_script = '/usr/bin/expect ./EXP-apreboot.exp'

#define sources for data from files
prefix = '/home/customer/russ/wicldumps/'
#prefix = ''
#prefix = '../wicldumps/'
fromfiles_provisioning = prefix + 'RAWDATA-provisioning.dat'
fromfiles_alarms = prefix + 'RAWDATA-alarms.dat'
fromfiles_fddcell = prefix + 'RAWDATA-fddcell.dat'

#set a foodfiles directory to scan for newest file
# must have the trailing '/' !
foodFilesDir = '/home/customer/russ/pm_foodfiles/'
#foodFilesDir = '../pm_foodfiles/'


#default log file name
defaultlogfilename = 'pm.log'

#define a long banner string when we run an OS command
#declaring here saves space in the rest of the code
rob = "=-=-=-=-=- RUNNING OS COMMAND =-=-=-=-=-: "


'''current list of active FGW's monitored by Operations. This info will 
change slowly over time and the code will need to be updated
This could be done better. Probably just pass in a CSV to the script. '''

# key legend
# 'selector' - This is the menu selector item that shows up in the list
# 'fgwName' - This is the hostname of the FGW
# 'fgwIP' - this is the static IP of the FGW
# 'bsgDump' - this is a holding place for the output of bsr_get_bsg_reg cmd
# 'monitoredStatus' - whether the FGW is monitored or not
# 'clusterId' - this is the ALU/BC cluster ID for the FGW

fgwList = [
    { 'selector' : '03' , 
      'fgwName' : 'CRTNTX35CR9S01', 
      'fgwDesc' : 'Carrollton 1' , 
      'fgwIP' : '10.24.40.171' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '101'},
    { 'selector' : '04' , 
      'fgwName' : 'SNTDCAUJCR9S01', 
      'fgwDesc' : 'Santa Clara 1' , 
      'fgwIP' : '107.96.253.139' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '1'},
    { 'selector' : '05' , 
      'fgwName' : 'SNAPCAHHCR9S01', 
      'fgwDesc' : 'Santa Ana 1' , 
      'fgwIP' : '10.170.70.75' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '51'},
    { 'selector' : '11' , 
      'fgwName' : 'SNAPCAHHCR9S02', 
      'fgwDesc' : 'Santa Ana 2' , 
      'fgwIP' : '10.170.70.77' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '52'},
    { 'selector' : '01' , 
      'fgwName' : 'NWTPMOABCR9S01', 
      'fgwDesc' : 'Hazelwood 1' , 
      'fgwIP' : '10.174.33.171' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '151'},
    { 'selector' : '02' , 
      'fgwName' : 'NWTPMOABCR9S02', 
      'fgwDesc' : 'Hazelwood 2' , 
      'fgwIP' : '10.174.33.173' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '152'},
    { 'selector' : '06' , 
      'fgwName' : 'RCPKNJ02CR9S01', 
      'fgwDesc' : 'Rochelle Park 1' , 
      'fgwIP' : '10.191.178.171' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '201'},
    { 'selector' : '07' , 
      'fgwName' : 'ATLNGAACCR9S01', 
      'fgwDesc' : 'Decatur 1' , 
      'fgwIP' : '10.187.167.139' , 
      'bsgDump' : '', 
      'monitoredStatus' : 
      'Monitored', 
      'clusterId' : '301'},
    { 'selector' : '08' , 
      'fgwName' : 'LKMRFLMFCR9S01', 
      'fgwDesc' : 'Lake Mary 1' , 
      'fgwIP' : '10.187.167.43' , 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '351'},
    { 'selector' : '09' , 
      'fgwName' : 'CLMAMDORCR9S01', 
      'fgwDesc' : 'Columbia 1' , 
      'fgwIP' : '10.188.131.203', 
      'bsgDump' : '', 
      'monitoredStatus' : 'Monitored', 
      'clusterId' : '251'},
    ]
'''
{ 'selector' : '10' , 
'fgwName' : 'CRTNTX35CR9S02', 
'fgwDesc' : 'Carrollton 2' , 
'fgwIP' : '10.24.40.173', 
'bsgDump' : '', 
'monitoredStatus' : 'Not Monitored', 
'clusterId' : '102'}
'''
    
    
#make the default ASCII escape code blank so text shows up normally
colorlist = [
    {'index' : '1', 
     'example' : '\033[1;31mRed like Radish\033[m', 
     'code' : '31m'},
    {'index' : '2', 
     'example' : '\033[1;32mGreen like Grass\033[m', 
     'code' : '32m'},
    {'index' : '3', 
     'example' : '\033[1;33mYellow like Yolk\033[m', 
     'code' : '33m'},
    {'index' : '4', 
     'example' : '\033[1;34mBlue like Blood\033[m', 
     'code' : '34m'},
    {'index' : '5', 
     'example' : '\033[1;35mMagenta like Mimosa\033[m', 
     'code' : '35m'},
    {'index' : '6', 
     'example' : '\033[1;36mCyan like Caribbean\033[m', 
     'code' : '36m'},
    {'index' : '7', 
     'example' : '\033[1;37mWhite like Whipped Cream\033[m', 
     'code' : '37m'},
    {'index' : '8',  
     'example' : '\033[1;38mCrimson like Chianti\033[m', 
     'code' : '38m'},
    {'index' : '9', 
     'example' : '\033[1;41mHighlighted Red like Radish\033[m', 
     'code' : '41m'},
    {'index' : '10', 
     'example' : '\033[1;42mHighlighted Green like Grass\033[m', 
     'code' : '42m'},
    {'index' : '11', 
     'example' : '\033[1;43mHighlighted Brown like Bear\033[m', 
     'code' : '43m'},
    {'index' : '12', 
     'example' : '\033[1;44mHighlighted Blue like Blood\033[m', 
     'code' : '44m'},
    {'index' : '13', 
     'example' : '\033[1;45mHighlighted Magenta like Mimosa\033[m', 
     'code' : '45m'},
    {'index' : '14', 
     'example' : '\033[1;46mHighlighted Cyan like Caribbean\033[m', 
     'code' : '46m'},
    {'index' : '15', 
     'example' : '\033[1;47mHighlighted Gray like Ghost\033[m', 
     'code' : '47m'},
    {'index' : '16', 
     'example' : '\033[1;48mHighlighted Crimson like Chianti\033[m', 
     'code' : '48m'},
    {'index' : '17', 
     'example' : 'None like Nancy', 
     'code': ''}
        ]
        
timeoutlist = [{'index' : '1', 'timeout' : '8'},
                {'index' : '2', 'timeout' : '12'},
                {'index' : '3', 'timeout' : '16'},
                {'index' : '4', 'timeout' : '24'},
                {'index' : '5', 'timeout' : '40'},
                {'index' : '6', 'timeout' : '50'},
        ]
##############################################################        
#Begin global section to define column headers and widths
##############################################################
'''              BEGIN ALARM DISPLAY FORMAT CONFIG                  '''
'''This section defines the formats for how Alarm objects are displayed '''
# this defines the Alarm output column widths and placement order
display_format_alarm = (
        '{0:7}'                    # index
        '{1:10}'                   # AlarmID
        '{2:31}'                   # specificProblems
        '{3:20}'                   # eventTime
        '{4:73}'                   # alarmedComponent
        '{5:20}'                   # bSRName
        '{6:14}'                   # monitoredStatus
        '{7:12}'                   # adminStatus
        '{8:4}'                    # clusterID
        '{9:15}'                   # fgwName
        )

# this is what's displayed at the bottom of the columns on screen output 
alarmheaderlist = ['index',                 # 0
                   'AlarmID',               # 1
                   'specificProblems',      # 2
                   'eventTime',             # 3
                   'alarmedComponent',      # 4
                   'bSRName',               # 5
                   'monitored?',            # 6
                   'adminStatus',           # 7
                   'cid',                   # 8
                   'fgwName']               # 9

# this is what's displayed at the top of the columns in the output Alarm csv
alarmcsvlist = ['AlarmID',
                'specificProblems',
                'eventTime',
                'alarmedComponent',
                'bSRName',
                'monitored?',
                'adminStatus',
                'cid',
                'fgwName']

#now turn the Alarm CSV header list into a comma separated string
header_alarm_csv = ','.join(alarmcsvlist)

'''              END ALARM DISPLAY FORMAT CONFIG                  '''
                
                
'''          BEGIN METROCELL DISPLAY FORMAT CONFIG                  '''                
'''This section defines the formats for how Metrocell objects are displayed '''
#this defines the 'short' format column widths and placement order
display_format_metrocell = (
        '{0:7}'             # index
        '{1:13}'            # bsrName
        '{2:11}'            # serial
        '{3:7}'             # bsrId
        '{4:23}'            # groupName
        '{5:31}'            # specInfo
        '{6:15}'            # monitoredStatus
        '{7:16}'            # ipsecIpAddr
        '{8:5}'             # clusterId
        '{9:5}'             # groupId
        '{10:15}'           # groupId
        '{11:9}'            # fgw
        '{12:12}'           # gagr
        '{13:6}'            # adminStatus
        )
#this defines the full format column widths and placement order
display_format_metrocell_full = (
        '{0:7}'             # index
        '{1:13}'            # bsrName
        '{2:11}'            # serial
        '{3:7}'             # bsrId
        '{4:23}'            # groupName
        '{5:23}'            # ctsEquipName
        '{6:31}'            # specInfo
        '{7:15}'            # monitoredStatus
        '{8:16}'            # ipsecIpAddr
        '{9:5}'             # clusterId
        '{10:5}'            # groupId
        '{11:15}'           # fgw
        '{12:9}'            # gagr
        '{13:9}'            # usid
        '{14:12}'           # faLocCode
        '{15:8}'            # inout
        '{16:5}'            # technology
        '{17:7}'            # alarm?
        '{18:4}'            # use case (UC)
        '{19:19}'           # swVersion
        '{20:12}'           # adminStatus
        )
# define column header strings for Metrocell/femto
# this is what's displayed at the bottom of the columns on output
femtoheaderlist = ['index',             # 0
                   'bsrName',           # 1
                   'serial',            # 2
                   'bsrId',             # 3
                   'groupName',         # 4
                   'specInfo',          # 5
                   'monitored?',        # 6
                   'ipsecIpAddr',       # 7
                   'clstr',             # 8
                   'grp',               # 9
                   'fgw',               # 10
                   'gagr',              # 11
                   'adminStatus',       # 12
                   'alarm?']            # 13

femtoheaderlist_full = [
                    'index',            # 0
                    'bsrName',          # 1
                    'serial',           # 2
                    'bsrId',            # 3
                    'groupName',        # 4
                    'ctsEquipName',     # 5
                    'specInfo',         # 6
                    'monitored?',       # 7
                    'ipsecIpAddr',      # 8
                    'clstr',            # 9
                    'grp',              # 10
                    'fgw',              # 11
                    'gagr',             # 12
                    'usid',             # 13
                    'faLocCode',        # 14
                    'inout',            # 15
                    'tech',             # 16
                    'alarm?',           # 17
                    'UC',               # 18
                    'swVersion',        # 19
                    'adminStatus'       # 20
                    ]

# what's displayed at the top of the columns in the SHORT output Metrocell csv
femtocsvlist = ['bsrName',
                'serial',
                'bsrId',
                'groupName',
                'specInfo',
                'monitored?',
                'ipsecIpAddr',
                'clusterId',
                'groupId',
                'fgw',
                'grid-gateway',
                'adminStatus',
                'bsrAdminStatus',
                'bsrOperStatus',
                'cellAdminStatus',
                'cellOperStatus',
                'UEconnected?'
                ]
# what's displayed at the top of the columns in the FULL output Metrocell csv
femtocsvlist_full = ['index',
                     'bsrName',
                     'serial',
                     'bsrId',
                     'groupName',
                     'ctsEquipName',
                     'specInfo',
                     'monitored?',
                     'ipsecIpAddr',
                     'clstr',
                     'grp',
                     'fgw',
                     'gagr',
                     'usid',
                     'faLocCode',
                     'inout',
                     'tech',
                     'alarm?',
                     'usecase',
                     'swVersion',
                     'adminStatus'
                     ]


display_format_metrocell_status = (
            '{0:17}'            # monitoredStatus
            '{1:16}'            # ipsecIpAddr
            '{2:9}'             # gagr
            '{3:12}'            # adminStatus
            '{4:12}'            # bsrAStatus
            '{5:12}'            # bsrOStatus
            '{6:13}'            # cellAStatus
            '{7:13}'            # cellOStatus
            '{8:5}'             # UEs?
            '{9:5}'             # clstr
            '{10:5}'            # grp
            '{11:15}'           # fgw
            '{12:7}'            # alarm?
            '{13:14}'           # bsrName
            '{14:19}'           # mac address
            '{15:16}'           # local ip addr
            '{16:16}'           # firstPublicIpInPath
            '{17:16}'           # defaultroute
            '{18:16}'           # netmask
            '{19:16}'           # dhcps ip
            )

statusheaderlist = [
        'monitoredStatus',      # 0
        'ipsecIpAddr',          # 1
        'gagr',                 # 2
        'adminStatus',          # 3
        'bsrAStatus',           # 4
        'bsrOStatus',           # 5
        'cellAStatus',          # 6
        'cellOStatus',          # 7
        'UEs?',                 # 8
        'clstr',                # 9
        'grp',                  # 10
        'fgw',                  # 11
        'alarm?',               # 12
        'bsrName',              # 13
        'macaddr',              # 14
        'lcl_ip',              # 15
        'firstpublic',              # 16
        'def_rt',              # 17
        'netmask',              # 18
        'dhcps_ip',              # 19
        ]

statuscsvlist = [
        'monitoredStatus',
        'ipsecIpAddr',
        'gagr',
        'adminStatus',
        'bsrAStatus',
        'bsrOStatus',
        'cellAStatus',
        'cellOStatus',
        'UEconnected?',
        'bsrName',
        'macaddr',
        'local_ip_addr',
        'firstPublicIpInPath',
        'default_route',
        'netmask',
        'dhcp_server_ip',
        ]

#now turn the Metrocell CSV header lists into comma separated strings
header_metro_csv = ','.join(femtocsvlist)
header_metro_full_csv = ','.join(femtocsvlist_full)
header_status_csv = ','.join(statuscsvlist)

'''          END METROCELL DISPLAY FORMAT CONFIG                  '''                


'''here we'll set up our variable length separators 
(e.g., --------) using the above column header strings'''
def genseparator(headerstring):
    sep = ''
    separator = sep.join(['-' for x in range(len(headerstring))])
    return separator

#now build the actual strings
header_status = display_format_metrocell_status.format(*statusheaderlist)
header_alarm = display_format_alarm.format(*alarmheaderlist)
header_metro = display_format_metrocell.format(*femtoheaderlist)
header_metrofull = display_format_metrocell_full.format(*femtoheaderlist_full)

sep_status = genseparator(header_status)
sep_alarm = genseparator(header_alarm)
sep_metro = genseparator(header_metro)
sep_metrofull = genseparator(header_metrofull)

'''End global section to define column headers and widths'''

#we'll build a menu pre-set list so UI is more fluid
preset_list = { '10': { 'columnName' : 'additionalText' },
                '11': { 'columnName' : 'alarmedComponent' },
                '12': { 'columnName' : 'notificationIdentifier' },
                '13': { 'columnName' : 'bSRName' },
                '14': { 'columnName' : 'FemtoCluster' },
                '15': { 'columnName' : 'eventTime' },
                '16': { 'columnName' : 'specificProblems' },
                '24': { 'attribute' : 'groupName' },
                '23': { 'attribute' : 'groupId' },
                '20': { 'attribute' : 'bsrName' },
                '21': { 'attribute' : 'bsrId' },
                '22': { 'attribute' : 'serial' },
                '29': { 'attribute' : 'clusterId' },
                '26': { 'attribute' : 'ipsecIpAddr' },
                '27': { 'attribute' : 'ctsname' },
                '28': { 'attribute' : 'usid' },
                '25': { 'attribute' : 'faloc' },
            }

autoupdate_frequency_list = [
    {'index' : '1', 
     'description' : '10 minutes (default)', 
     'time' : 600},
    {'index' : '2', 
     'description' : '20 minutes', 
     'time' : 1200},
    {'index' : '3', 
     'description' : '5 minutes', 
     'time' : 300},
    {'index' : '999', 
     'description' : '10 seconds (WARNING: Debug only)', 
     'time' : 10}, 
        ]
            
# sometimes we want to dump FGW BSG Registrations output to file for offline
#   parsing. I'll define filename prefix here:
bsgdump_fname_prefix = 'DATA_BSGDUMP_'
bsgdump_bool = False
#bsgdump_offline_pull_dir = './'
bsgdump_offline_pull_dir = '../wicldumps/'

